source/extensions/filters/listener/tls_inspector/tls_inspector.cc

```diff
@@ -0,0 +1,198 @@
+}
+
+void Filter::onServername(absl::string_view name) {
+  if (!name.empty()) {
+    config_->stats().sni_found_.inc();
+    cb_->socket().setRequestedServerName(name);
+  } else {
+    config_->stats().no_sni_found_.inc();
+  }
+  clienthello_success_ = true;
+
+  // TODO(ggreenway): Notify the ConnectionSocket that this
+  // is a TLS connection.
+}
+
+void Filter::onRead() {
+  // This receive code is somewhat complicated, because it must be done as a MSG_PEEK because
+  // there is no way for a listener-filter to pass payload data to the ConnectionImpl and filters
+  // that get created later.
+  //
+  // The file_event_ in this class gets events everytime new data is available on the socket,
+  // even if previous data has not been read, which is always the case due to MSG_PEEK. When
+  // the TlsInspector completes and passes the socket along, a new FileEvent is created for the
+  // socket, so that new event is immediately signalled as readable because it is new and the socket
+  // is readable, even though no new events have ocurred.
+  //
+  // TODO(ggreenway): write an integration test to ensure the events work as expected on all
+  // platforms.
+  auto& os_syscalls = Api::OsSysCallsSingleton::get();
+  ssize_t n = os_syscalls.recv(cb_->socket().fd(), buf_.data(), buf_.size(), MSG_PEEK);
+  ENVOY_LOG(trace, "tls inspector: recv: {}", n);
+
+  if (n == -1 && errno == EAGAIN) {
+    return;
+  } else if (n < 0) {
+    config_->stats().read_error_.inc();
+    done(false);
+    return;
+  }
+
+  // Because we're doing a MSG_PEEK, data we've seen before gets returned every time, so
+  // skip over what we've already processed.
+  if (static_cast<uint64_t>(n) > read_) {
+    const uint8_t* data = buf_.data() + read_;
+    const size_t len = n - read_;
+    read_ = n;
+    parseClientHello(data, len);
+  }
+}
+
+void Filter::onTimeout() {
+  ENVOY_LOG(trace, "tls inspector: timeout");
+  config_->stats().read_timeout_.inc();
+  done(false);
+}
+
+void Filter::done(bool success) {
+  ENVOY_LOG(trace, "tls inspector: done: {}", success);
+  timer_.reset();
+  file_event_.reset();
+  cb_->continueFilterChain(success);
+}
+
+void Filter::parseClientHello(const void* data, size_t len) {
+  // Ownership is passed to ssl_ in SSL_set_bio()
+  BIO* bio = BIO_new_mem_buf(data, len);
+  SSL_set_bio(ssl_.get(), bio, bio);
+
+  // Make the mem-BIO return that there is more data
+  // available beyond it's end
+  BIO_set_mem_eof_return(bio, -1);
+
+  int ret = SSL_do_handshake(ssl_.get());
+
+  switch (SSL_get_error(ssl_.get(), ret)) {
+  case SSL_ERROR_WANT_READ:
+    if (read_ == buf_.size()) {
+      // We've hit the specified size limit. This is an unreasonably large ClientHello;
+      // indicate failure.
+      config_->stats().client_hello_too_big_.inc();
+      done(false);
+    }
+    break;
+  case SSL_ERROR_SSL:
+    if (clienthello_success_) {
+      done(true);
+    } else {
+      config_->stats().invalid_client_hello_.inc();
```

Comment:
Yes. This stat is really "either not TLS at all, or TLS with an invalid ClientHello".  The difference between the two is really a matter of how the heuristic is tuned for whether it's TLS or not. Is there a better name I can give this stat, that isn't a full sentence?

Reply:
Maybe `tls_found` and `no_tls_found`, to match detected protocol and SNI stats?

While technically this detects invalid ClientHello, it's an edge case, and majority of the time this will actually mean it's valid non-TLS traffic and not malformed ClientHello.